test/dns-test.js

```diff
@@ -0,0 +1,396 @@
+  for (let i = 0; i < count; i++) {
+    const obj = { complete: false };
+    node.once('block', () => {
+      obj.complete = true;
+    });
+    await nclient.execute('generatetoaddress', [1, address]);
+    await common.forValue(obj, 'complete', true);
+  }
+}
+
+/**
+ * Verify DNSSEC for each name in the
+ * response. Only check the answer and
+ * authority sections. If all rrsets
+ * are signed, the responses will be
+ * too large and not fit into udp packets
+ */
+
+function verifyDNSSEC(resource, pubkey, qtype, name) {
+  // rr types that are not committed to in DNSSEC
+  const skip = new Set([
+    types.RRSIG
+  ]);
+
+  name = util.fqdn(name);
+
+  const {answer, authority} = resource;
+  const targets = [answer, authority];
+
+  const records = [];
+
+  for (const target of targets) {
+    const toVerify = new Set();
+    for (const record of target) {
+      records.push(record);
+
+      if (!skip.has(record.type))
+        toVerify.add(record.type);
+    }
+
+    for (const type of toVerify) {
+      const rrsig = target.find((rr) => {
+        return rr.type === types.RRSIG
+          && rr.data.typeCovered === type;
+      });
+
+      if (name !== rrsig.name) {
+        const record = records.find((rr) => {
+          return rr.name === util.fqdn(name)
+            && (rr.type === types.CNAME
+            || rr.type === types.DNAME
+            || rr.type === types.SRV
+            || rr.type === types.NS);
+        });
+
+        assert(record);
+
+        switch (qtype) {
+          case types.CNAME:
+          case types.DNAME:
+          case types.SRV:
+            name = record.data.target;
+            break;
+          case types.NS:
+            name = record.data.ns;
+            break;
+          default:
+            assert(false, 'rrsig name does not match');
+        }
+      }
+
+      const rrs = util.extractSet(target, name, type);
+
+      const valid = dnssec.verify(rrsig, pubkey, rrs);
+      assert(valid);
+    }
+  }
+}
+
+/**
+ * Nullify out any signatures in
+ * a DNS response.
+ */
+
+function nullSig(resource) {
+  for (const answer of resource.answer) {
+    if (answer.data.signature)
+      answer.data.signature = null;
+  }
+
+  for (const authority of resource.authority) {
+    if (authority.data.signature)
+      authority.data.signature = null;
+  }
+
+  for (const additional of resource.additional) {
+    if (additional.data.signature)
+      additional.data.signature = null;
+  }
+}
+
+/**
+ * Parse the ZSK and KSK from
+ * a record.
+ */
+
+function getSigningKeys(record) {
+  let ksk, zsk;
+  for (const rr of record.answer) {
+    if (rr.type === types.DNSKEY) {
+      const {data} = rr.toJSON();
+      if (data.keyType === 'ZSK')
+        zsk = rr;
+      else if (data.keyType === 'KSK')
+        ksk = rr;
+    }
+  }
+
+  return {
+    zsk: zsk,
+    ksk: ksk
+  };
+}
+
+/**
+ * Build a name to query for the types
+ * that require additional labels.
```

Comment:
I'd like to turn this into a helper function as I think I'd find it useful building applications on top of Handshake. Not quite sure where it would live or how to generalize it correctly - notice that it uses a magic index into an array.

Reply:
I'd like to check through `hsd` and `bns` to make sure that it already doesn't exist. Would the fn signature be:

`name (str), hsd resource (Resource), dns type (int) -> List[str] (list of query strings)`

A Resource can have multiple query strings depending on the Records inside of it, but this also may not return the name number of strings in the return value as the number of Records in the Resource which may be confusing to a library consumer.

Reply:
OK if we not maybe `test/util/common` is a good place.

Reply:
I think we should leave this as a specific test helper for now. We can abstract this into a proper library method when the functionality is needed elsewhere.

Reply:
Ok this makes sense to me. Its not clearly documented anywhere in the Handshake docs that you need to use a particular domain name (with subdomains) to query for some of the rr types. Maybe the `bns` stub resolver could be extended for this repo where its wrapped with an easier to use API that will handle these cases as well as use handshake native types instead of dns rr types. I could imagine that being very helpful for onboarding new developers, especially if it works in the browser.